arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -0,0 +1,464 @@
+#define XLAT_TABLE_SIZE_SHIFT   PAGE_SIZE_SHIFT /* Size of one complete table */
+#define XLAT_TABLE_SIZE		(1U << XLAT_TABLE_SIZE_SHIFT)
+
+#define XLAT_TABLE_ENTRY_SIZE_SHIFT	3U /* Each table entry is 8 bytes */
+#define XLAT_TABLE_LEVEL_MAX	3U
+
+#define XLAT_TABLE_ENTRIES_SHIFT \
+			(XLAT_TABLE_SIZE_SHIFT - XLAT_TABLE_ENTRY_SIZE_SHIFT)
+#define XLAT_TABLE_ENTRIES	(1U << XLAT_TABLE_ENTRIES_SHIFT)
+
+/* Address size covered by each entry at given translation table level */
+#define L3_XLAT_VA_SIZE_SHIFT	PAGE_SIZE_SHIFT
+#define L2_XLAT_VA_SIZE_SHIFT	\
+			(L3_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+#define L1_XLAT_VA_SIZE_SHIFT	\
+			(L2_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+#define L0_XLAT_VA_SIZE_SHIFT	\
+			(L1_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+
+#define LEVEL_TO_VA_SIZE_SHIFT(level) \
+				(PAGE_SIZE_SHIFT + (XLAT_TABLE_ENTRIES_SHIFT * \
+				(XLAT_TABLE_LEVEL_MAX - (level))))
+
+/* Virtual Address Index within given translation table level */
+#define XLAT_TABLE_VA_IDX(va_addr, level) \
+	((va_addr >> LEVEL_TO_VA_SIZE_SHIFT(level)) & 0x1ffULL)
+
+/*
+ * Calculate the initial translation table level from CONGIG_ARM64_VA_BITS
+ * For a 4 KB page size,
+ * (va_bits <= 21)	 - base level 3
+ * (22 <= va_bits <= 30) - base level 2
+ * (31 <= va_bits <= 39) - base level 1
+ * (40 <= va_bits <= 47) - base level 0
+ */
+#define GET_XLAT_TABLE_BASE_LEVEL(va_bits)	\
+	((va_bits > L0_XLAT_VA_SIZE_SHIFT)	\
+	? 0U					\
+	: (va_bits > L1_XLAT_VA_SIZE_SHIFT)	\
+	? 1U					\
+	: (va_bits > L2_XLAT_VA_SIZE_SHIFT)	\
+	? 2U : 3U)
+
+#define XLAT_TABLE_BASE_LEVEL	GET_XLAT_TABLE_BASE_LEVEL(CONFIG_ARM64_VA_BITS)
+
+#define GET_NUM_BASE_LEVEL_ENTRIES(va_bits)	\
+	(1U << (va_bits - LEVEL_TO_VA_SIZE_SHIFT(XLAT_TABLE_BASE_LEVEL)))
+
+#define NUM_BASE_LEVEL_ENTRIES	GET_NUM_BASE_LEVEL_ENTRIES(CONFIG_ARM64_VA_BITS)
+
+#if DUMP_PTE
+#define L0_SPACE ""
+#define L1_SPACE "  "
+#define L2_SPACE "    "
+#define L3_SPACE "      "
+#define XLAT_TABLE_LEVEL_SPACE(level)		\
+	(((level) == 0) ? L0_SPACE :		\
+	((level) == 1) ? L1_SPACE :		\
+	((level) == 2) ? L2_SPACE : L3_SPACE)
+#endif
+
+static u64_t base_xlat_table[NUM_BASE_LEVEL_ENTRIES]
+		__aligned(NUM_BASE_LEVEL_ENTRIES * sizeof(u64_t));
+
+static u64_t xlat_tables[CONFIG_MAX_XLAT_TABLES][XLAT_TABLE_ENTRIES]
+		__aligned(XLAT_TABLE_SIZE);
+
+/* Translation table control register settings */
+static u64_t get_tcr(int el)
+{
+	u64_t tcr;
+	u64_t pa_bits = CONFIG_ARM64_PA_BITS;
+	u64_t va_bits = CONFIG_ARM64_VA_BITS;
+	u64_t tcr_ps_bits;
+
+	switch (pa_bits) {
+	case 48:
+		tcr_ps_bits = TCR_PS_BITS_256TB;
+		break;
+	case 44:
+		tcr_ps_bits = TCR_PS_BITS_16TB;
+		break;
+	case 42:
+		tcr_ps_bits = TCR_PS_BITS_4TB;
+		break;
+	case 40:
+		tcr_ps_bits = TCR_PS_BITS_1TB;
+		break;
+	case 36:
+		tcr_ps_bits = TCR_PS_BITS_64GB;
+		break;
+	default:
+		tcr_ps_bits = TCR_PS_BITS_4GB;
+		break;
+	}
+
+	if (el == 1) {
+		tcr = TCR_EL1_RSVD | (tcr_ps_bits << TCR_EL1_IPS_SHIFT);
```

Comment:
What `TCR_EL1_RSVD` is for? Is this implementation defined?

Reply:
will look into this and get back.

Reply:
In `TCR_EL1`, bit `[15:0]` is for memory region addressed `TTBR0_EL1 `and bit `[31:16]` is for memory region addressed `TTBR1_EL1`.

TCR_EL1_RSVD  is `(1 << 30)`. It used to indicate translation granule along with bit 30 (`TG1[31:30]`). It is defined for some cores and reserved for some. So, yes, this field is implementation defined.

However, we have disabled translation table walk using `TTBR1_EL1`, so this field should not matter run time.

Reference: ARMv8-A ARM (DDI 0487E.a): D13.2.112 `TCR_EL1`, Translation Control Register (EL1)

Reply:
Yes, this is what I thought. Being something custom (implementation defined) I would probably leave that out.

Reply:
Not touching implementation defined bits in TCR.